pkg/timeutil/periodic.go

```diff
@@ -0,0 +1,240 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package timeutil
+
+import (
+	"fmt"
+	"strings"
+	"time"
+)
+
+type periodicStart struct {
+	dayOfWeek      int
+	hourOfDay      int
+	minuteOfHour   int
+	secondOfMinute int
+}
+
+type Periodic struct {
+	start    periodicStart
+	duration time.Duration
+}
+
+type Period struct {
+	Start time.Time
+	End   time.Time
+}
+
+// ParsePeriodic returns a Periodic specified as a start and duration.
+func ParsePeriodic(start, duration string) (pc Periodic, err error) {
+	if pc.start, err = parseStart(start); err != nil {
+		return pc, fmt.Errorf("unable to parse start: %v", err)
+	}
+	if pc.duration, err = time.ParseDuration(duration); err != nil {
+		return pc, fmt.Errorf("unable to parse duration: %v", err)
+	}
+	if pc.duration < time.Duration(0) {
+		return pc, fmt.Errorf("duration cannot be negative")
+	}
+	return
+}
+
+var weekdays = map[string]int{
+	"sun": int(time.Sunday),
+	"mon": int(time.Monday),
+	"tue": int(time.Tuesday),
+	"wed": int(time.Wednesday),
+	"thu": int(time.Thursday),
+	"fri": int(time.Friday),
+	"sat": int(time.Saturday),
+}
+
+// parseStart parses a string into a periodicStart.
+func parseStart(start string) (ps periodicStart, err error) {
+	ps.dayOfWeek = -1
+	f := strings.Fields(start)
+	nf := len(f)
+
+	if nf != 1 && nf != 2 {
+		return ps, fmt.Errorf("wrong number of fields")
+	}
+
+	if nf == 2 {
+		if dow, ok := weekdays[strings.ToLower(f[0])]; ok {
```

Comment:
I'm a switch junkie, imho things like this become a lot cleaner and easier to maintain
```
switch len(f) {
  case 1: // no day provided
  case 2:
     ps.dayOfWeek, ok := weekdays[strings.ToLower(f[0])]
     if !ok {
        return nil, fmt.Errorf("invalid day of week: %q", f[0])
     }
  default:
     return nil, fmt.Errorf("wrong  number of fields")
}
```